### PR TITLE
[Bugfix] Fix duplicate \n and </think> in ThinkingBudgetLogitProcessor due to overlap scheduling race condition

### PR DESCRIPTION
**Pull request by magicllm**
## Motivation

When using `ThinkingBudgetLogitProcessor` (e.g., `Qwen3ThinkingBudgetLogitProcessor`) with
`thinking_budget` set to a small value, the generated output contains duplicate `\n` and
duplicate `</think>` tokens.

Expected: `\n</think>` (one newline, one end-of-thinking tag)
Actual: `\n\n</think></think>` (duplicated)

## Root Cause
SGLang uses two threads in overlap scheduling mode:
- **Thread 1 (CPU scheduling thread)**: appends the sampled token to `req.output_ids`
- **Thread 2 (GPU computation thread)**: runs `CustomLogitProcessor`, which reads `req.output_ids`

Due to the overlap, when Thread 2 executes step N+1, Thread 1 may not have committed step N's
token to `output_ids` yet. This is a **known race condition**.

In `ThinkingBudgetLogitProcessor`, this race condition causes two specific problems:
1. **Duplicate `\n`**: At step N, the processor forces `\n` (because budget is exceeded and
   `output_ids[-1] != \n`). At step N+1, `output_ids` still doesn't contain the `\n` from
   step N, so the processor forces `\n` again.
2. **Duplicate `</think>`**: Same mechanism — the forced `</think>` from step N hasn't been
   written to `output_ids` when step N+1 reads it, so `</think>` is forced again.

## Modifications

Store one-shot state flags in `param_dict` (which is only accessed by the GPU thread, so no
race condition) to track the forcing steps:
- **`_forcing_newline`**: Set when `\n` is forced. Next step reads and pops this flag, skips
  reading `output_ids[-1]`, and directly forces `</think>`.
- **`_forcing_end`**: Set when `</think>` is forced. Next step reads and pops this flag, and
  blocks `</think>` for that single step to prevent duplication.
Both flags are consumed via `dict.pop()` (read + delete in one operation), so they only take
effect for exactly **one step** — just enough to cover the race condition window. After that,
the original `cur_ids`-based logic takes over normally. No tokens are permanently blocked.

## Accuracy Tests

We have verified the effectiveness of this bug fix on Qwen3-30B-A3B-Thinking-2507, Qwen3-235B-A22B-Thinking-2507, Baichuan-M3, Baichuan-M2

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/sampling/custom_logit_processor.py [hunk 1]
```diff
@@ -80,6 +80,17 @@ def __call__(self, logits, custom_param_list: list[dict[str, Any]]):
...
+
+            if param_dict.pop("_forcing_end", False):
+                logits[i, self.THINKING_END_TOKEN_ID] = -float("inf")
+                continue
+
+            if param_dict.pop("_forcing_newline", False):
+                logits[i, :] = -float("inf")
+                logits[i, self.THINKING_END_TOKEN_ID] = 0.0
+                param_dict["_forcing_end"] = True
+                continue
```
[the remark this answers is not in the document]
Comment (magicllm) [reply python/sglang/srt/sampling/custom_logit_processor.py]: The code has been modified to improve the maintainability.
